util/versionutils.c

```diff
@@ -0,0 +1,333 @@
+    return (-5); // undefined
+
+  if (part1 == NULL)
+    {
+      g_free (part2);
+      if (rs2)
+        return (rs2 - release_state1);
+      else
+        return (-1);
+    }
+
+  if (part2 == NULL)
+    {
+      g_free (part1);
+      if (rs1)
+        return (release_state2 - rs1);
+      else
+        return (1);
+    }
+
+  int ret = -5;
+
+  if (rs1 && rs2)
+    ret = rs2 - rs1;
+
+  if (rs1)
+    ret = -1;
+
+  if (rs2)
+    ret = 1;
+
+  if (!rs1 && !rs2 && atoi (part1) < atoi (part2))
+    ret = -1;
+
+  if (!rs1 && !rs2 && atoi (part1) == atoi (part2))
+    ret = 0;
+
+  if (!rs1 && !rs2 && atoi (part1) > atoi (part2))
+    ret = 1;
+
+  g_free (part1);
+  g_free (part2);
+  g_free (ver1);
+  g_free (ver2);
+  return (ret);
+}
+
+/**
+ * @brief Prepare the version string for comparison.
+ *
+ * @param[in]  version  The version string to generate the prepared
+ *                      version string from.
+ *
+ * @return  Returns a prepared copy of the version string version.
+ */
+static char *
+prepare_version_string (const char *version)
+{
+  char prep_version[1024];
+  char *ver;
+  int index_v, index_pv;
+  gboolean is_digit;
+
+  if (!version)
+    return (NULL);
+
+  ver = g_strdup (version);
+
+  /* set all characters to lowercase */
+  char *c = ver;
+  for (; *c; c++)
+    *c = tolower (*c);
+
+  index_v = index_pv = 0;
+
+  is_digit = g_ascii_isdigit (ver[0]);
+
+  while (index_v < (int) strlen (ver))
+    {
+      if (ver[index_v] == '_' || ver[index_v] == '-' || ver[index_v] == '+'
+          || ver[index_v] == ':' || ver[index_v] == '.')
+        {
+          if (index_pv > 0 && prep_version[index_pv - 1] != '.')
+            {
+              prep_version[index_pv] = '.';
+              index_pv++;
+            }
+          index_v++;
+          continue;
+        }
+
+      if (is_digit != g_ascii_isdigit (ver[index_v]))
+        {
+          is_digit = !is_digit;
+          if (index_pv > 0 && prep_version[index_pv - 1] != '.')
+            {
+              prep_version[index_pv] = '.';
+              index_pv++;
+            }
+        }
+
+      if (ver[index_v] == 'r')
+        {
+          if (strstr (ver + index_v, "releasecandidate") == ver + index_v)
+            {
+              prep_version[index_pv] = 'r';
+              prep_version[index_pv + 1] = 'c';
+              index_pv += 2;
+              index_v += 16;
+              continue;
+            }
+          if ((strstr (ver + index_v, "release-candidate") == ver + index_v)
+              || (strstr (ver + index_v, "release_candidate") == ver + index_v))
+            {
+              prep_version[index_pv] = 'r';
+              prep_version[index_pv + 1] = 'c';
+              index_pv += 2;
+              index_v += 17;
+              continue;
+            }
+        }
+
+      prep_version[index_pv] = ver[index_v];
+      index_v++;
+      index_pv++;
+    }
+
+  prep_version[index_pv] = '\0';
+  g_free (ver);
+  return (g_strdup (prep_version));
```

Comment:
You're using g_strdup so I think you should return a gchar*. I know that in the past we aimed to keep gchar* out of the headers, but many of the newer headers do use gchar*.

Also I recommend dropping the parenthesis around all your return values, to match all the other code.